node/src/main/scala/io/casperlabs/node/NodeRuntime.scala

```
@@ -74,130 +75,133 @@ class NodeRuntime private[node] (
-                              executionEngineService,
-                              _.value
-                            )(
-                              labEff,
-                              Metrics.eitherT(Monad[Task], metrics),
-                              blockStore,
-                              Cell.eitherTCell(Monad[Task], rpConnections),
-                              NodeDiscovery.eitherTNodeDiscovery(Monad[Task], nodeDiscovery),
-                              TransportLayer.eitherTTransportLayer(Monad[Task], log, transport),
-                              ErrorHandler[Effect],
-                              eitherTrpConfAsk(rpConfAsk),
-                              oracle,
-                              Capture[Effect],
-                              Concurrent[Effect],
-                              Time.eitherTTime(Monad[Task], time),
-                              Log.eitherTLog(Monad[Task], log),
-                              multiParentCasperRef,
-                              blockDagStorage,
-                              scheduler
-                            )
-    packetHandler = PacketHandler.pf[Effect](casperPacketHandler.handle)(
-      Applicative[Effect],
-      Log.eitherTLog(Monad[Task], log),
-      ErrorHandler[Effect]
-    )
-    nodeCoreMetrics = diagnostics.effects.nodeCoreMetrics[Task]
-    jvmMetrics      = diagnostics.effects.jvmMetrics[Task]
-
-    program = nodeProgram[Effect](executionEngineService)(
-      Monad[Effect],
-      time,
-      rpConfState,
-      rpConfAsk,
-      peerNodeAsk,
-      metrics,
-      transport,
-      kademliaRPC,
-      nodeDiscovery,
-      rpConnections,
-      blockStore,
-      oracle,
-      packetHandler,
-      multiParentCasperRef,
-      nodeCoreMetrics,
-      jvmMetrics
-    )
-    _ <- handleUnrecoverableErrors(program)
-  } yield ()
+
+  val main = GrpcExecutionEngineService[Effect](
+    conf.grpcServer.socket,
+    conf.server.maxMessageSize
+  ).use(runMain)
```

So how does this work exactly during shutdown? The `nodeProgram` starts a few loops:
```scala
      _       <- NodeDiscovery[Task].discover.attemptAndLog.executeOn(loopScheduler).start.toEffect
      _       <- Task.defer(casperLoop.forever.value).executeOn(loopScheduler).start.toEffect
      _       <- EitherT(Task.defer(loop.forever.value).executeOn(loopScheduler))
```
I guess the first two are running asynchronously, and the last one will basically never complete the deferred, so the thing will never return. We also use `addShutdownHook` to register a callback that closes gRPC servers and databases. I guess that during shutdown the monix run loop will stop scheduling tasks, right, so it will stop the Fibers we started as well as the `time.sleep(1.minute)` in loop. 

My question is, will it actually execute the `release` part of the resource? Is it a graceful cancelation of averything inside with a nice `ExitCase.Cancel` that gives a chance for those to run?

---

Yes, it will, I have checked it manually.
As I understand `IOApp` trait will force `release` on all introduced brackets on termination, more details can be revealed by exploring `IOAppPlatform`, but I didn't dive into it. 
I think it just uses that nice property of `Bracket` when it guarantees that all sub-brackets will `release` when release called on the parent.

---

Awesome, cheers!

---

There is a lot of details and corner cases to `cats`' bracket, its release and acquisition phases. I don't know all of them but it doesn't call `release` when there is an error happening in the `acquire` phase (or sth like that). It was widely discussed on gitter and twitter recently.

---

A couple of weeks ago I played with `Task.bracket` to understand its behaviour: https://scastie.scala-lang.org/igor-ramazanov/mJot8javSAyDxmIc8dmJyw

---

>I don't know all of them but it doesn't call `release` when there is an error happening in the `acquire` phase (or sth like that). It was widely discussed on gitter and twitter recently.

I think it's reasonable because:
1) we can not release something that hasn't been acquired
2) we want to fail fast

---

On gitter they didn't reach consensus on that.

---

Or there was sth about interrupting the _forever running_ task :thinking: can't remember exactly.